From c19665b2dcd5b53fb2284cd0e6a9a4e31e5d4e93 Mon Sep 17 00:00:00 2001
From: Alexandru Costache <alexandru@balena.io>
Date: Mon, 18 Jan 2021 17:17:46 +0100
Subject: [PATCH] u-boot: Integrate with BalenaOS, load kernel from rootfs

Many patches have been added to get the board
boot BalenaOS. We now have squashed all of them
into one commit, so that we can keep the repository
a bit more manageable.

Apart from the existing ones, we've now also switched to
loading the kernel from the current active rootfs. We also
load the dtb from the rootfs, because we've noticed
4.9 kernel is incompatible with 4.14 dtb, and we can
prevent failed boot this way in the future.

Upstream-status: Inappropriate [configuration]
Signed-off-by: Alexandru Costache <alexandru@balena.io>
---
 Kconfig                                       |  5 ++++
 .../imx8mq_phanbell/imx8m_phanbell.c          | 17 -----------
 common/env_common.c                           | 11 +++++++
 configs/mx8mq_phanbell_defconfig              |  5 ++++
 drivers/usb/gadget/f_fastboot.c               | 29 -------------------
 include/configs/imx8mq_evk_android.h          |  8 -----
 include/configs/imx8mq_evk_androidthings.h    |  7 -----
 include/configs/imx8mq_phanbell.h             | 28 +++++++++---------
 include/configs/imx8qm_arm2_android.h         |  8 -----
 include/configs/imx8qm_mek_android.h          |  8 -----
 include/configs/imx8qxp_arm2_android.h        |  8 -----
 include/configs/imx8qxp_mek_android.h         |  7 -----
 include/configs/mx_android_common.h           |  3 --
 13 files changed, 36 insertions(+), 108 deletions(-)

diff --git a/Kconfig b/Kconfig
index 02f954d64e..0240d72faa 100644
--- a/Kconfig
+++ b/Kconfig
@@ -325,6 +325,11 @@ config ARCH_FIXUP_FDT_MEMORY
 	  used for booting OS with different memory setup where the part of
 	  the memory location should be used for different purpose.
 
+config CMD_FS_UUID
+        bool "fsuuid command"
+        help
+          Enables fsuuid command for filesystem UUID.
+
 endmenu		# Boot images
 
 source "api/Kconfig"
diff --git a/board/freescale/imx8mq_phanbell/imx8m_phanbell.c b/board/freescale/imx8mq_phanbell/imx8m_phanbell.c
index b066ed2a5e..110f5083c1 100644
--- a/board/freescale/imx8mq_phanbell/imx8m_phanbell.c
+++ b/board/freescale/imx8mq_phanbell/imx8m_phanbell.c
@@ -332,23 +332,6 @@ int board_late_init(void)
 	setenv("board_rev", "iMX8MQ");
 #endif
 
-/* If we aren't supporting Android, set the default boot command
- * to run boot.scr from partition one of the boot device.
- */
-#ifdef CONFIG_SD_BOOT
-	static char bootdev[32];
-	static char bootcmd[128];
-	snprintf(bootdev, sizeof(bootdev), "%d", mmc_dev);
-	setenv("bootdev", bootdev);
-	/* For Zircon, try booting from mmc if ext2load fails */
-	snprintf(bootcmd, sizeof(bootcmd), "ext2load mmc %d:1 ${loadaddr} boot.scr; source; boota mmc0 boot_a;", mmc_dev);
-	setenv("bootcmd", bootcmd);
-#endif
-
-#ifdef CONFIG_ENV_IS_IN_MMC
-	board_late_mmc_env_init();
-#endif
-
 	return 0;
 }
 
diff --git a/common/env_common.c b/common/env_common.c
index 7fb62e8b45..e10292c150 100644
--- a/common/env_common.c
+++ b/common/env_common.c
@@ -194,6 +194,17 @@ int env_import(const char *buf, int check)
 	env_t *ep = (env_t *)buf;
 	int ret;
 
+       /* Setting CONFIG_ENV_IS_NOWHERE triggers compilation
+        * failure for setups that have the fastboot patches.
+        * Disabling fastboot config triggers others failures.
+        *
+        * Simply force using BalenaOS environment, the one embedded
+        * in the boot binary, and leave the eMMC one intact.
+        */
+       set_default_env("Switch to BalenaOS environment... \n");
+
+       return 0;
+
 	if (check) {
 		uint32_t crc;
 
diff --git a/configs/mx8mq_phanbell_defconfig b/configs/mx8mq_phanbell_defconfig
index 74a6fdcc14..f5d18b7ee4 100644
--- a/configs/mx8mq_phanbell_defconfig
+++ b/configs/mx8mq_phanbell_defconfig
@@ -42,3 +42,8 @@ CONFIG_ANDROID_BOOT_IMAGE=y
 CONFIG_ZIRCON_BOOT_IMAGE=y
 CONFIG_SD_BOOT=y
 CONFIG_CMD_IMPORTENV=y
+CONFIG_CMD_PART=y
+CONFIG_CMD_EXPORTENV=y
+CONFIG_FAT_WRITE=y
+CONFIG_CMD_UNZIP=y
+CONFIG_CMD_FS_UUID=y
diff --git a/drivers/usb/gadget/f_fastboot.c b/drivers/usb/gadget/f_fastboot.c
index 42f1959fb3..18b2b2fb22 100755
--- a/drivers/usb/gadget/f_fastboot.c
+++ b/drivers/usb/gadget/f_fastboot.c
@@ -1186,35 +1186,6 @@ unsigned int fastboot_flash_get_ptn_count(void)
 #ifdef CONFIG_FSL_FASTBOOT
 void board_fastboot_setup(void)
 {
-#if defined(CONFIG_FASTBOOT_STORAGE_MMC)
-	static char boot_dev_part[32];
-	u32 dev_no;
-#endif
-	switch (get_boot_device()) {
-#if defined(CONFIG_FASTBOOT_STORAGE_MMC)
-	case SD1_BOOT:
-	case SD2_BOOT:
-	case SD3_BOOT:
-	case SD4_BOOT:
-	case MMC1_BOOT:
-	case MMC2_BOOT:
-	case MMC3_BOOT:
-	case MMC4_BOOT:
-	case USB_BOOT:
-		dev_no = mmc_get_env_dev();
-		sprintf(boot_dev_part,"mmc%d",dev_no);
-		if (!getenv("fastboot_dev"))
-			setenv("fastboot_dev", boot_dev_part);
-		sprintf(boot_dev_part, "boota mmc%d", dev_no);
-		if (!getenv("bootcmd"))
-			setenv("bootcmd", boot_dev_part);
-		break;
-#endif /*CONFIG_FASTBOOT_STORAGE_MMC*/
-	default:
-		printf("unsupported boot devices\n");
-		break;
-	}
-
 	/* add soc type into bootargs */
 	if (is_mx6dqp()) {
 		if (!getenv("soc_type"))
diff --git a/include/configs/imx8mq_evk_android.h b/include/configs/imx8mq_evk_android.h
index ea1902940d..48b73a2f5e 100644
--- a/include/configs/imx8mq_evk_android.h
+++ b/include/configs/imx8mq_evk_android.h
@@ -43,12 +43,4 @@
 #define CONFIG_SUPPORT_RAW_INITRD
 #define CONFIG_SERIAL_TAG
 
-#undef CONFIG_EXTRA_ENV_SETTINGS
-#undef CONFIG_BOOTCOMMAND
-
-#define CONFIG_EXTRA_ENV_SETTINGS		\
-	"splashpos=m,m\0"			\
-	"fdt_high=0xffffffffffffffff\0"		\
-	"initrd_high=0xffffffffffffffff\0"	\
-
 #endif /* IMX8MQ_EVK_ANDROID_H */
diff --git a/include/configs/imx8mq_evk_androidthings.h b/include/configs/imx8mq_evk_androidthings.h
index cf68349d7b..66d90b5070 100644
--- a/include/configs/imx8mq_evk_androidthings.h
+++ b/include/configs/imx8mq_evk_androidthings.h
@@ -42,12 +42,5 @@
 #define CONFIG_SUPPORT_RAW_INITRD
 #define CONFIG_SERIAL_TAG
 
-#undef CONFIG_EXTRA_ENV_SETTINGS
-#undef CONFIG_BOOTCOMMAND
-
-#define CONFIG_EXTRA_ENV_SETTINGS		\
-	"splashpos=m,m\0"			\
-	"fdt_high=0xffffffffffffffff\0"		\
-	"initrd_high=0xffffffffffffffff\0"	\
 
 #endif /* IMX8MQ_EVK_ANDROID_THINGS_H */
diff --git a/include/configs/imx8mq_phanbell.h b/include/configs/imx8mq_phanbell.h
index f903060a4c..006b1dda8a 100644
--- a/include/configs/imx8mq_phanbell.h
+++ b/include/configs/imx8mq_phanbell.h
@@ -70,7 +70,7 @@
 /* Flat Device Tree Definitions */
 #define CONFIG_OF_BOARD_SETUP
 
-#undef CONFIG_CMD_EXPORTENV
+
 #undef CONFIG_CMD_IMLS
 
 #undef CONFIG_CMD_CRC32
@@ -112,24 +112,25 @@
 #define CONFIG_EXTRA_ENV_SETTINGS		\
 	CONFIG_MFG_ENV_SETTINGS \
 	"script=boot.scr\0" \
-	"image=Image\0" \
+	"image=Image.gz\0" \
 	"console=ttymxc0,115200 earlycon=ec_imx6q,0x30860000,115200\0" \
 	"fdt_addr=0x43000000\0"			\
 	"fdt_high=0xffffffffffffffff\0"		\
 	"boot_fdt=try\0" \
 	"fdt_file=fsl-imx8mq-phanbell.dtb\0" \
 	"initrd_addr=0x43800000\0"		\
+	"zip_addr=0x43800000\0" \
 	"initrd_high=0xffffffffffffffff\0" \
 	"mmcdev="__stringify(CONFIG_SYS_MMC_ENV_DEV)"\0" \
 	"mmcpart=" __stringify(CONFIG_SYS_MMC_IMG_LOAD_PART) "\0" \
 	"mmcroot=" CONFIG_MMCROOT " rootwait rw\0" \
 	"mmcautodetect=yes\0" \
-	"mmcargs=setenv bootargs console=${console} root=${mmcroot}\0 " \
+	"mmcargs=setenv bootargs console=${console} ${resin_kernel_root} ro rootwait ${os_cmdline} net.ifnames=0\0" \
 	"loadbootscript=fatload mmc ${mmcdev}:${mmcpart} ${loadaddr} ${script};\0" \
 	"bootscript=echo Running bootscript from mmc ...; " \
 		"source\0" \
-	"loadimage=fatload mmc ${mmcdev}:${mmcpart} ${loadaddr} ${image}\0" \
-	"loadfdt=fatload mmc ${mmcdev}:${mmcpart} ${fdt_addr} ${fdt_file}\0" \
+	"loadimage=ext4load mmc ${mmcdev}:${resin_root_part} ${zip_addr} boot/${image}; unzip ${zip_addr} ${loadaddr};\0" \
+	"loadfdt=ext4load mmc ${mmcdev}:${resin_root_part} ${fdt_addr} boot/${fdt_file}\0" \
 	"mmcboot=echo Booting from mmc ...; " \
 		"run mmcargs; " \
 		"if test ${boot_fdt} = yes || test ${boot_fdt} = try; then " \
@@ -163,14 +164,14 @@
 		"fi;\0"
 
 #define CONFIG_BOOTCOMMAND \
+           "setenv resin_kernel_load_addr ${loadaddr};" \
+           "run resin_set_kernel_root; run set_os_cmdline;" \
+           "setenv mmcdev ${resin_dev_index};" \
+           "setenv mmcpart ${resin_root_part};" \
 	   "mmc dev ${mmcdev}; if mmc rescan; then " \
-		   "if run loadbootscript; then " \
-			   "run bootscript; " \
-		   "else " \
-			   "if run loadimage; then " \
-				   "run mmcboot; " \
-			   "else run netboot; " \
-			   "fi; " \
+		   "if run loadimage; then " \
+			   "run mmcboot; " \
+		   "else run netboot; " \
 		   "fi; " \
 	   "else booti ${loadaddr} - ${fdt_addr}; fi"
 
@@ -189,7 +190,7 @@
 
 #define CONFIG_ENV_OVERWRITE
 #define CONFIG_ENV_OFFSET               (64 * SZ_64K)
-#define CONFIG_ENV_SIZE			0x1000
+#define CONFIG_ENV_SIZE			0x4000
 #define CONFIG_ENV_IS_IN_MMC
 #define CONFIG_SYS_MMC_ENV_DEV		0   /* USDHC2 */
 #define CONFIG_MMCROOT			"/dev/mmcblk1p2"  /* USDHC2 */
@@ -233,6 +234,7 @@
 #define CONFIG_CMD_EXT4
 #define CONFIG_CMD_EXT4_WRITE
 #define CONFIG_CMD_FAT
+#define CONFIG_FAT_WRITE
 
 #define CONFIG_SUPPORT_EMMC_BOOT	/* eMMC specific */
 #define CONFIG_SYS_MMC_IMG_LOAD_PART	1
diff --git a/include/configs/imx8qm_arm2_android.h b/include/configs/imx8qm_arm2_android.h
index b2bf785eb0..32c829625c 100644
--- a/include/configs/imx8qm_arm2_android.h
+++ b/include/configs/imx8qm_arm2_android.h
@@ -49,14 +49,6 @@
 #define CONFIG_SUPPORT_RAW_INITRD
 #define CONFIG_SERIAL_TAG
 
-#undef CONFIG_EXTRA_ENV_SETTINGS
-#undef CONFIG_BOOTCOMMAND
-
-#define CONFIG_EXTRA_ENV_SETTINGS					\
-	"splashpos=m,m\0"	  \
-	"fdt_high=0xffffffffffffffff\0"	  \
-	"initrd_high=0xffffffffffffffff\0" \
-
 #define CONFIG_FASTBOOT_BUF_ADDR   CONFIG_SYS_LOAD_ADDR
 #define CONFIG_FASTBOOT_BUF_SIZE   0x19000000
 
diff --git a/include/configs/imx8qm_mek_android.h b/include/configs/imx8qm_mek_android.h
index 63c226e0e7..811eb31147 100644
--- a/include/configs/imx8qm_mek_android.h
+++ b/include/configs/imx8qm_mek_android.h
@@ -55,14 +55,6 @@
 #define CONFIG_SUPPORT_RAW_INITRD
 #define CONFIG_SERIAL_TAG
 
-#undef CONFIG_EXTRA_ENV_SETTINGS
-#undef CONFIG_BOOTCOMMAND
-
-#define CONFIG_EXTRA_ENV_SETTINGS					\
-	"splashpos=m,m\0"	  \
-	"fdt_high=0xffffffffffffffff\0"	  \
-	"initrd_high=0xffffffffffffffff\0" \
-
 #define CONFIG_FASTBOOT_BUF_ADDR   CONFIG_SYS_LOAD_ADDR
 #define CONFIG_FASTBOOT_BUF_SIZE   0x19000000
 
diff --git a/include/configs/imx8qxp_arm2_android.h b/include/configs/imx8qxp_arm2_android.h
index cc98c72c6b..abe9d191c7 100644
--- a/include/configs/imx8qxp_arm2_android.h
+++ b/include/configs/imx8qxp_arm2_android.h
@@ -49,14 +49,6 @@
 #define CONFIG_SUPPORT_RAW_INITRD
 #define CONFIG_SERIAL_TAG
 
-#undef CONFIG_EXTRA_ENV_SETTINGS
-#undef CONFIG_BOOTCOMMAND
-
-#define CONFIG_EXTRA_ENV_SETTINGS					\
-	"splashpos=m,m\0"	  \
-	"fdt_high=0xffffffffffffffff\0"	  \
-	"initrd_high=0xffffffffffffffff\0" \
-
 #define CONFIG_FASTBOOT_BUF_ADDR   CONFIG_SYS_LOAD_ADDR
 #define CONFIG_FASTBOOT_BUF_SIZE   0x19000000
 
diff --git a/include/configs/imx8qxp_mek_android.h b/include/configs/imx8qxp_mek_android.h
index 02a22462a4..44c8fcf409 100644
--- a/include/configs/imx8qxp_mek_android.h
+++ b/include/configs/imx8qxp_mek_android.h
@@ -51,13 +51,6 @@
 #define CONFIG_SUPPORT_RAW_INITRD
 #define CONFIG_SERIAL_TAG
 
-#undef CONFIG_EXTRA_ENV_SETTINGS
-#undef CONFIG_BOOTCOMMAND
-
-#define CONFIG_EXTRA_ENV_SETTINGS					\
-	"splashpos=m,m\0"	  \
-	"fdt_high=0xffffffffffffffff\0"	  \
-	"initrd_high=0xffffffffffffffff\0" \
 
 #define CONFIG_FASTBOOT_BUF_ADDR   CONFIG_SYS_LOAD_ADDR
 #define CONFIG_FASTBOOT_BUF_SIZE   0x19000000
diff --git a/include/configs/mx_android_common.h b/include/configs/mx_android_common.h
index 6a00ef229e..18f6858cbf 100644
--- a/include/configs/mx_android_common.h
+++ b/include/configs/mx_android_common.h
@@ -40,9 +40,6 @@
 #define CONFIG_SUPPORT_RAW_INITRD
 #define CONFIG_SERIAL_TAG
 
-#undef CONFIG_EXTRA_ENV_SETTINGS
-#undef CONFIG_BOOTCOMMAND
-
 #define CONFIG_EXTRA_ENV_SETTINGS	\
 	"splashpos=m,m\0"	\
 	"fdt_high=0xffffffff\0"	\
-- 
2.17.1

